src/mito2/src/compaction/run.rs

```diff
@@ -37,12 +33,68 @@ pub(crate) trait Ranged {
     {
         let (lhs_start, lhs_end) = self.range();
         let (rhs_start, rhs_end) = other.range();
-        match lhs_start.cmp(&rhs_start) {
-            Ordering::Less => lhs_end >= rhs_start,
-            Ordering::Equal => true,
-            Ordering::Greater => lhs_start <= rhs_end,
+
+        lhs_start.max(rhs_start) < lhs_end.min(rhs_end)
+    }
+}
+
+pub fn find_overlapping_items<T: Ranged + Clone>(l: &mut [T], r: &mut [T]) -> Vec<T> {
+    if l.is_empty() || r.is_empty() {
+        return vec![];
+    }
+
+    let mut res = Vec::with_capacity(l.len() + r.len());
+
+    // Sort both arrays by start boundary for more efficient overlap detection
+    sort_ranged_items(l);
+    sort_ranged_items(r);
```

Comment:
Consider checking if l and r are already sorted to avoid sorting them again in `reduce_run()`. Maybe we can add a sorted flag in T.

Reply:
fixed in https://github.com/GreptimeTeam/greptimedb/pull/6121/commits/067a6bd82279cd794a7bc18346ab8ea62d4050c2